read-fonts/src/tables/postscript/stack.rs

```diff
@@ -0,0 +1,381 @@
+    values: [i32; MAX_STACK],
+    value_is_fixed: [bool; MAX_STACK],
+    top: usize,
+}
+
+impl Stack {
+    pub fn new() -> Self {
+        Self {
+            values: [0; MAX_STACK],
+            value_is_fixed: [false; MAX_STACK],
+            top: 0,
+        }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.top == 0
+    }
+
+    pub fn len(&self) -> usize {
+        self.top
+    }
+
+    pub fn verify_exact_len(&self, len: usize) -> Result<(), Error> {
+        if self.top != len {
+            Err(Error::StackUnderflow)
+        } else {
+            Ok(())
+        }
+    }
+
+    pub fn verify_at_least_len(&self, len: usize) -> Result<(), Error> {
+        if self.top < len {
+            Err(Error::StackUnderflow)
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Returns true if the number of elements on the stack is odd.
+    ///
+    /// Used for processing some charstring operators where an odd
+    /// count represents the presence of the glyph advance width at the
+    /// bottom of the stack.
+    pub fn len_is_odd(&self) -> bool {
+        self.top & 1 != 0
+    }
+
+    pub fn clear(&mut self) {
+        self.top = 0;
+    }
+
+    /// Reverse the order of all elements on the stack.
+    ///
+    /// Some charstring operators are simpler to process on a reversed
+    /// stack.
+    pub fn reverse(&mut self) {
+        self.values[..self.top].reverse();
+        self.value_is_fixed[..self.top].reverse();
+    }
+
+    pub fn push_i32(&mut self, value: i32) -> Result<(), Error> {
+        self.push(value, false)
+    }
+
+    pub fn push_fixed(&mut self, value: Fixed) -> Result<(), Error> {
+        self.push(value.to_bits(), true)
+    }
+
+    /// Returns the 32-bit integer at the given index on the stack.
+    ///
+    /// Will return an error if the value at that index was not pushed as an
+    /// integer.
+    pub fn get_i32(&self, index: usize) -> Result<i32, Error> {
+        let value = *self
+            .values
+            .get(index)
+            .ok_or(Error::InvalidStackAccess(index))?;
+        if self.value_is_fixed[index] {
+            // FreeType returns an error here rather than converting
+            // <https://gitlab.freedesktop.org/freetype/freetype/-/blob/80a507a6b8e3d2906ad2c8ba69329bd2fb2a85ef/src/psaux/psstack.c#L145>
+            Err(Error::ExpectedI32StackEntry(index))
+        } else {
+            Ok(value)
+        }
+    }
+
+    /// Returns the 16.16 fixed point value at the given index on the stack.
+    ///
+    /// If the value was pushed as an integer, it will be automatically
+    /// converted to 16.16 fixed point.
+    pub fn get_fixed(&self, index: usize) -> Result<Fixed, Error> {
+        let value = *self
+            .values
+            .get(index)
+            .ok_or(Error::InvalidStackAccess(index))?;
+        Ok(if self.value_is_fixed[index] {
+            Fixed::from_bits(value)
+        } else {
+            Fixed::from_i32(value)
+        })
```

Comment:
this is funny, what's with the asymmetry?

Reply:
If you mean the fact that pop_int requires an integer but pop_fixed does a conversion, the only answer I can give is “that’s what FreeType does.” In some ways, it makes sense— when we’re asking for an integer from the stack, it’s almost always used as an offset or index in which case we generally want the full 32 bits.